src/main/scala/tsiHost/TsiHostWidget.scala

```diff
@@ -0,0 +1,315 @@
+package testchipip
+
+import chisel3._
+import chisel3.util._
+import chisel3.core.{withClockAndReset}
+
+import freechips.rocketchip.config.{Parameters, Field}
+import freechips.rocketchip.subsystem.{BaseSubsystem}
+import freechips.rocketchip.regmapper.{HasRegMap, RegField}
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.tilelink._
+import freechips.rocketchip.util._
+
+import SerialAdapter._
+
+/**
+ * Parameters for the SerDes
+ *
+ * @param clientParams client parameters that receives the SerialIO inwards
+ * @param managerParams manager parameters that sends the SerialIO outwards
+ */
+case class TSIHostSerdesParams(
+  clientParams: TLClientParameters = TLClientParameters(
+    name = "tl-tsi-host-serdes",
+    sourceId = IdRange(0, 1)),
+  managerParams: TLManagerParameters = TLManagerParameters(
+    address = Seq(AddressSet(0, BigInt("FFFFFFFF", 16))),
+    regionType = RegionType.UNCACHED,
+    supportsGet = TransferSizes(1, 64),
+    supportsPutFull = TransferSizes(1, 64),
+    supportsPutPartial = TransferSizes(1, 64))
+)
+
+
+/**
+ * TSI Host parameter class
+ *
+ * @param mmioRegWidth size of the MMIO data being sent back and forth in bits (connects to SerialAdapter)
+ * @param serialIfWidth size of the serialIO *out* of the widget (connects TLSerdessers)
+ * @param txQueueEntries size of the queue for sending TSI requests
+ * @param rxQueueEntries size of the queue for receiving TSI responses
+ * @param baseAddress start address of the MMIO registers
+ */
+case class TSIHostParams(
+  mmioRegWidth: Int = 32,
+  serialIfWidth: Int = 32,
+  txQueueEntries: Int = 16,
+  rxQueueEntries: Int = 16,
+  baseAddress: BigInt = BigInt(0x10017000),
+  serdesParams: TSIHostSerdesParams = TSIHostSerdesParams()
+)
+
+/**
+ * Offsets for the MMIO communication queues (base + offset to get the proper address)
+ */
+object TSIHostWidgetCtrlRegs {
+  val txQueueOffset = 0x00 // note: these assume mmioRegWidth = 32b
```

Comment:
Currently this is `required` indirectly in the backend by the statement `require(params.mmioRegWidth == SERIAL_IF_WIDTH)` (where the SERIAL_IF_WIDTH) is 32. I can add another require in the frontend for this also if you'd like.

Reply:
IMO we shouldn't expose mmioRegWidth as a param at all.

Reply:
This depends on the future of the widget and FESVR. If we want more BW we can send a 64b data value through the system. This would give a little bit more speed but would require changes to this widget, and to the serialadapter. If you don't think that'll happen anytime soon, we can just hardcode that.

Reply:
Note: There can be some confusion between `serialIfWidth` and `SERIAL_IF_WIDTH`. The former is the interface with going out/into the serdes while the latter is the interface with from the MMIO to the serial adapter.

Reply:
IMO `SERIAL_IF_WIDTH` should be `XLen` but I think that change is outside the scope of this PR. I'm not opposed to leaving `mmioRegWidth` as long as there's an assertion. Also yes, `SERIAL_IF_WIDTH` is named poorly, but that's not your fault :). I think `serialIfWidth` is the appropriate name for what it does.